### PR TITLE
Retirement bonds - KI Integration

### PR DESCRIPTION
**Pull request by cujowolf**
Contracts for both the underlying retirement bond contract to provide swaps on KI and direct retirements using carbon pool tokens. Included is an allocator contract to toggle and interface with the Treasury to fund and close markets.

Update `LibSwap` on the aggregator to take advantage of RB if possible, otherwise failover to performing the swap through existing DEX pools.

### CODE REVIEW

File: src/protocol/bonds/CarbonRetirementBondDepository.sol [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+ * @notice A smart contract that handles the distribution of carbon in exchange for KLIMA tokens.
+ * Bond depositors can only use this to retire carbon by providing KLIMA tokens.
+ */
+
+contract CarbonRetirementBondDepository is Ownable2Step {
+    using SafeERC20 for IKlima;
+
+    /// @notice Address of the KLIMA token contract.
+    address public constant KLIMA = 0x4e78011Ce80ee02d2c3e649Fb657E45898257815;
+    /// @notice Address of the DAO contract.
```
Comment (ShimonD-KlimaDAO) [inline src/protocol/bonds/CarbonRetirementBondDepository.sol:22]: typo - DAO wallet

File: src/protocol/bonds/CarbonRetirementBondDepository.sol [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    address public constant DAO = 0x65A5076C0BA74e5f3e069995dc3DAB9D197d995c;
+    /// @notice Address of the Treasury contract.
+    address public constant TREASURY = 0x7Dd4f0B986F032A44F913BF92c9e8b7c17D77aD7;
+    /// @notice address of the Klima Infinity contract.
+    address public constant INFINITY = 0x8cE54d9625371fb2a068986d32C85De8E6e995f8;
+    /// @notice Divisor used for calculating percentages.
+    uint256 public constant FEE_DIVISOR = 10000;
```
Comment (ShimonD-KlimaDAO) [inline src/protocol/bonds/CarbonRetirementBondDepository.sol:29]: 10_000 would be better for readability

File: src/infinity/libraries/TokenSwap/LibSwap.sol [hunk 3]
```diff
@@ -167,6 +193,55 @@ library LibSwap {
...
+        return _performToExactSwap(0, C.sushiRouter(), path, sourceAmount, klimaAmount);
+    }
+
+    function swapWithRetirementBonds(
+        address sourceToken,
+        address carbonToken,
+        uint256 sourceAmount,
+        uint256 carbonAmount
+    ) internal returns (uint256 carbonRecieved) {
+        if (sourceToken == C.klima()) {
```
Comment (ShimonD-KlimaDAO) [inline src/infinity/libraries/TokenSwap/LibSwap.sol:224]: I suggest also adding support for sKlima and wsKlima here out of the gate.

**Review by ShimonD-KlimaDAO — CHANGES_REQUESTED** [review]
Looks good overall, but would be nice to make a few small updates, esecially the sklima/wsklima support.

File: src/infinity/libraries/TokenSwap/LibSwap.sol (hunk 3 above)
Comment (cujowolf) [reply src/infinity/libraries/TokenSwap/LibSwap.sol:224]: If the source token is sKLIMA or wsKLIMA it gets changed to KLIMA on line 39.

**Review by ShimonD-KlimaDAO — APPROVED** [review]
Looks good!

**Comment by cujowolf** [comment]
resolves #40

**Pull request merged**